xontrib/hist_navigator.py

```diff
@@ -74,6 +85,26 @@ def listd():
     print(XSH_DIRS_HISTORY.history)
 
 
+def _p_msg_fmt(s):
+    return tokenize_ansi(PygmentsTokens(partial_color_tokenize(XSH.shell.shell.prompt_formatter(s)))) # noqa
+
+def _update_prompt(): # force-update all 3 prompts (if exist) to new values
+    # Helpful when xontrib changes prompt vars in the background and
+    # doesn't want to wait for the next prompt cycle
+    shellx = XSH.shell.shell
+    shellx.prompt_formatter.fields.reset() # reset fields cache
+    if prompt_l := envx['PROMPT']:
+        prompt_l = prompt_l() if callable(prompt_l) else prompt_l
+        shellx.prompter.message        = _p_msg_fmt(prompt_l)
+    if prompt_r := envx['RIGHT_PROMPT']:
+        prompt_r = prompt_r() if callable(prompt_r) else prompt_r
+        shellx.prompter.rprompt        = _p_msg_fmt(prompt_r)
+    if prompt_b := envx['BOTTOM_TOOLBAR']:
+        prompt_b = prompt_b() if callable(prompt_b) else prompt_b
+        shellx.prompter.bottom_toolbar = _p_msg_fmt(prompt_b)
+    shellx.prompter.app.invalidate()      # send signal that prompt needs update
```

Comment:
This only works for ptk-shell. It is better to check the shell type and fallback to old behaviour if inplace update of prompt is not possible

Reply:
is there a way to check just once on xontrib load and register respective class/functions
This fails since at the loading time I get `best` `SHELL_TYPE` instead of `prompt_toolkit`, so I guess it's actually chosen some time after the xontrib is loaded

If not, I'll move the check to inside the _move function

```py
class _DirsHistory:
    if envx.get('SHELL_TYPE') in ["prompt_toolkit", "prompt_toolkit2"]:
        is_ptk = True
    else:
        is_ptk = False
```

Reply:
meanwhile moved everything inside, on invokation the check seems fine

Reply:
wait, just tried testing with a readline shell and your xontrib keybinds don't seem to work there, think due to `on_ptk_create`, so keybinds are only registered in ptk shells, so at least the keybinding checks don't seem to make sense, will remove them

Reply:
yes that is good